Sources/AsyncHTTPClient/NIOTransportServices/TLSConfiguration.swift

```diff
@@ -0,0 +1,127 @@
+        switch self {
+        case .tlsv1:
+            return .TLSv10
+        case .tlsv11:
+            return .TLSv11
+        case .tlsv12:
+            return .TLSv12
+        case .tlsv13:
+            return .TLSv13
+        }
+    }
+}
+
+@available (macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+internal extension TLSConfiguration {
+    
+    /// Dispatch queue used by Network framework TLS to control certificate verification
+    static var tlsDispatchQueue = DispatchQueue(label: "TLSDispatch")
+
+    /// create NWProtocolTLS.Options for use with NIOTransportServices from the NIOSSL TLSConfiguration
+    ///
+    /// - Parameter queue: Dispatch queue to run `sec_protocol_options_set_verify_block` on.
+    /// - Returns: Equivalent NWProtocolTLS Options
+    func getNWProtocolTLSOptions() -> NWProtocolTLS.Options {
+        let options = NWProtocolTLS.Options()
+        
+        // minimum TLS protocol
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+            sec_protocol_options_set_min_tls_protocol_version(options.securityProtocolOptions, self.minimumTLSVersion.nwTLSProtocolVersion)
+        }
+        
+        // maximum TLS protocol
+        if let maximumTLSVersion = self.maximumTLSVersion {
+            if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+                sec_protocol_options_set_max_tls_protocol_version(options.securityProtocolOptions, maximumTLSVersion.nwTLSProtocolVersion)
+            } else {
+                precondition(self.maximumTLSVersion != nil, "TLSConfiguration.maximumTLSVersion is not supported")
+            }
+        }
+
+        // application protocols
+        if self.applicationProtocols.count > 0 {
+            preconditionFailure("TLSConfiguration.applicationProtocols is not supported")
+        }
+        /*for applicationProtocol in self.applicationProtocols {
+            applicationProtocol.utf8.withContiguousStorageIfAvailable { buffer in
+                guard let opaquePointer = OpaquePointer(buffer.baseAddress) else { return }
+                let int8Pointer = UnsafePointer<Int8>(opaquePointer)
+                sec_protocol_options_add_tls_application_protocol(options.securityProtocolOptions, int8Pointer)
+            }
+        }*/
+
+        // the certificate chain
+        if self.certificateChain.count > 0 {
+            preconditionFailure("TLSConfiguration.certificateChain is not supported")
+        }
+        
+        // cipher suites
+        if self.cipherSuites.count > 0 {
+            //preconditionFailure("TLSConfiguration.cipherSuites is not supported")
+        }
+        
+        // key log callback
+        if let _ = self.keyLogCallback {
+            preconditionFailure("TLSConfiguration.keyLogCallback is not supported")
+        }
+        
+        // private key
+        if let _ = self.privateKey {
+            preconditionFailure("TLSConfiguration.privateKey is not supported")
+        }
```

Comment:
Sadly, supporting this and `certificateChain` in a cross-platform way is really hard, I think. NWProtocolTLS wants a `sec_identity_t`. I know of no way to create a `sec_identity_t` from data in memory: it requires adding the private key to the keychain. We could try to do that along with a label that allowed us to try to find the same private key again to avoid duplicating it, but it's a lot of work. Leaving it for another day seems reasonable.